Signal/src/ViewControllers/ExperienceUpgradesPageViewController.swift

```diff
@@ -4,6 +4,160 @@
+        imageView.autoSetDimension(.height, toSize: ScaleFromIPhone5(225))
+
+        // Title label layout
+        titleLabel.autoSetDimension(.height, toSize: ScaleFromIPhone5(40))
+        titleLabel.autoPinWidthToSuperview(withMargin: ScaleFromIPhone5To7Plus(16, 24))
+        titleLabel.autoPinEdge(toSuperviewEdge: .top)
+
+        // Body label layout
+        bodyLabel.autoPinEdge(.top, to: .bottom, of: imageView, withOffset: ScaleFromIPhone5To7Plus(18, 28))
+        bodyLabel.autoPinWidthToSuperview(withMargin: bodyMargin)
+        bodyLabel.setContentHuggingVerticalHigh()
+
+        // Button layout
+        button.autoPinEdge(.top, to: .bottom, of: bodyLabel, withOffset: ScaleFromIPhone5(16))
+        button.autoPinWidthToSuperview(withMargin: ScaleFromIPhone5(32))
+
+        bottomSpacer.autoPinEdge(.top, to: .bottom, of: button, withOffset: ScaleFromIPhone5(16))
+        bottomSpacer.autoPinEdge(toSuperviewEdge: .bottom)
+        bottomSpacer.autoPinWidthToSuperview()
+    }
+
+    // MARK: - Actions
+
+    func addButton(title: String, action: @escaping (UIButton) -> Void) -> UIButton {
+        self.buttonAction = action
+        let button = MultiLineButton()
+        view.addSubview(button)
+        button.setTitle(title, for: .normal)
+        button.setTitleColor(UIColor.ows_signalBrandBlue(), for: .normal)
+        button.isUserInteractionEnabled = true
+        button.addTarget(self, action:#selector(didTapButton), for: .touchUpInside)
+        button.contentEdgeInsets = UIEdgeInsets(top: 10, left: 10, bottom: 10, right: 10)
+        button.titleLabel?.textAlignment = .center
+        button.titleLabel?.font = UIFont.ows_mediumFont(withSize: ScaleFromIPhone5(18))
+        return button
+    }
+
+    func didTapButton(sender: UIButton) {
+        Logger.debug("\(TAG) in \(#function)")
+
+        guard let buttonAction = self.buttonAction else {
+            owsFail("\(TAG) button action was nil")
+            return
+        }
+
+        buttonAction(sender)
+    }
+}
+
+/**
+ * Allows multiple lines of button text, and ensures the buttons intrinsic content size reflects that of it's label.
+ */
+class MultiLineButton: UIButton {
+
+    // MARK: - Init
+
+    required init?(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
+
+        self.commonInit()
+    }
+
+    required init() {
+        super.init(frame: CGRect.zero)
+
+        self.commonInit()
+    }
+
+    private func commonInit() {
+        self.titleLabel?.numberOfLines = 0
+        self.titleLabel?.lineBreakMode = .byWordWrapping
+    }
+
+    // MARK: - Overrides
+
+    override var intrinsicContentSize: CGSize {
+        guard let titleLabel = titleLabel else {
+            return CGSize.zero
+        }
+
+        // be more forgiving with the tappable area
+        let extraPadding: CGFloat = 20
+        let labelSize = titleLabel.intrinsicContentSize
```

Comment:
Given that the title label's contents can line wrap, don't we need to measure its size in light of a max width constraint?